### ISSUE TITLE
Assertion failure regcomp.c:14701: S_regatom: Assertion `FOLD' failed.

### ISSUE THREAD

**Issue by dur-randir** [Perl/perl5#17196]
This is a bug report for perl from afl@dorothy,
generated with the help of perlbug 1.41 running under perl 5.31.5.

-----------------------------------------------------------------
[Please describe your issue here]

While fuzzing perl v5.29.10-23-g7c0d7520a3 built with afl and run
under libdislocator, I found the following program

```
0=~'(?l)0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000'
```

to cause an assertion failure

perl: regcomp.c:14701: S_regatom: Assertion `FOLD' failed.

GDB stack trace is following

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7c24535 in __GI_abort () at abort.c:79
#2  0x00007ffff7c2440f in __assert_fail_base (fmt=0x7ffff7d86ee0 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x555555adf778 "FOLD",
    file=0x555555ad4490 "regcomp.c", line=14701, function=<optimized out>) at assert.c:92
#3  0x00007ffff7c32102 in __GI___assert_fail (assertion=0x555555adf778 "FOLD", file=0x555555ad4490 "regcomp.c", line=14701,
    function=0x555555af79d0 <__PRETTY_FUNCTION__.23909> "S_regatom") at assert.c:101
#4  0x00005555556d03d3 in S_regatom (pRExC_state=0x7fffffffd5f0, flagp=0x7fffffffcc94, depth=4) at regcomp.c:14701
#5  0x00005555556c0d4a in S_regpiece (pRExC_state=0x7fffffffd5f0, flagp=0x7fffffffcdb0, depth=3) at regcomp.c:12548
#6  0x00005555556c0661 in S_regbranch (pRExC_state=0x7fffffffd5f0, flagp=0x7fffffffce58, first=1, depth=2) at regcomp.c:12468
#7  0x00005555556bde77 in S_reg (pRExC_state=0x7fffffffd5f0, paren=0, flagp=0x7fffffffd308, depth=1) at regcomp.c:12170
#8  0x00005555556a0c57 in Perl_re_op_compile (patternp=0x0, pat_count=1, expr=0x555555c0d748, eng=0x555555bd5d20 <PL_core_reg_engine>, old_re=0x0,
    is_bare_re=0x0, orig_rx_flags=0, pm_flags=0) at regcomp.c:7737
#9  0x00005555555ba4a3 in Perl_pmruntime (o=0x555555c0d6e0, expr=0x555555c0d748, repl=0x0, flags=0, floor=0) at op.c:7445
#10 0x00005555555b326c in Perl_bind_match (type=30, left=0x555555c0d780, right=0x555555c0d748) at op.c:5436
#11 0x000055555566fdc1 in Perl_yyparse (gramtype=258) at perly.y:1048
#12 0x00005555555ecda0 in S_parse_body (env=0x0, xsinit=0x5555555a120f <xs_init>) at perl.c:2527
#13 0x00005555555eb072 in perl_parse (my_perl=0x555555be0260, xsinit=0x5555555a120f <xs_init>, argc=2, argv=0x7fffffffe1e8, env=0x0) at perl.c:1818
#14 0x00005555555a114d in main (argc=2, argv=0x7fffffffe1e8, env=0x7fffffffe200) at perlmain.c:132
```

This is a regression in blead, bisect points to

```
3363c7035ff1df0c3ffeae0cd18bb86cc39d62e4 is the first bad commit
commit 3363c7035ff1df0c3ffeae0cd18bb86cc39d62e4
Author: Karl Williamson <khw@cpan.org>
Date:   Thu Sep 26 21:38:46 2019 -0600

    regex: Create and handle LEXACT nodes

    See the previous commit for info on these.

    I am not changing trie code to recognize these at this time.
```

[Please do not change anything below this line]
-----------------------------------------------------------------
---
Flags:
    category=core
    severity=high
---
Site configuration information for perl 5.31.5:

Configured by root at Fri Oct 18 05:50:54 MSK 2019.

Summary of my perl5 (revision 5 version 31 subversion 5) configuration:
  Derived from: 859b78b1a8287681e524286f4e989b01816433ba
  Platform:
    osname=linux
    osvers=4.19.0-6-amd64
    archname=x86_64-linux
    uname='linux dorothy 4.19.0-6-amd64 #1 smp debian 4.19.67-2+deb10u1 (2019-09-20) x86_64 gnulinux '
    config_args='-des -Dusedevel -Dcc=gcc -DDEBUGGING -Doptimize=-O0 -g -ggdb3'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='gcc'
    ccflags ='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O0 -g -ggdb3'
    cppflags='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='8.3.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='gcc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.28.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.28'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O0 -g -ggdb3 -L/usr/local/lib -fstack-protector-strong'

Characteristics of this binary (from libperl):
  Compile-time options:
    DEBUGGING
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Locally applied patches:
    uncommitted-changes
  Built under linux
  Compiled at Oct 18 2019 06:02:50
  %ENV:
    PERLBREW_BASHRC_VERSION="0.78"
    PERLBREW_HOME="/home/afl/.perlbrew"
    PERLBREW_MANPATH="/home/afl/perlbrew/perls/perl-5.20.2/man"
    PERLBREW_PATH="/home/afl/perlbrew/bin:/home/afl/perlbrew/perls/perl-5.20.2/bin"
    PERLBREW_PERL="perl-5.20.2"
    PERLBREW_ROOT="/home/afl/perlbrew"
    PERLBREW_VERSION="0.78"
  @INC:
    lib
    /usr/local/lib/perl5/site_perl/5.31.5/x86_64-linux
    /usr/local/lib/perl5/site_perl/5.31.5
    /usr/local/lib/perl5/5.31.5/x86_64-linux
    /usr/local/lib/perl5/5.31.5

**Comment by atoomic** [comment]
bisect points to this commit 3363c7035ff1df0c3ffeae0cd18bb86cc39d62e4

**Comment by toddr** [comment]
@dur-randir can this be closed now f58502b2d60aee3f0a89514476499437ba6d8e52 is in blead?

**Comment by dur-randir** [comment]
Yes, fuzzer doesn't see it any more.
